pinot-broker/src/main/java/org/apache/pinot/broker/requesthandler/MultipleOrEqualitiesToInClauseFilterQueryTreeOptimizer.java

```diff
@@ -20,148 +20,291 @@
 import java.util.Map;
 import java.util.Set;
-import java.util.TreeSet;
 import org.apache.pinot.common.request.FilterOperator;
 import org.apache.pinot.common.utils.request.FilterQueryTree;
 
 
 /**
- * Optimizer that collapses multiple OR clauses to IN clauses. For example, <code>a = 1 OR a = 2 OR a =
- * 3</code> gets turned to <code>a IN (1, 2, 3)</code>.
+ * The optimizer looks at the FilterQueryTree to convert
+ *
+ * (1) multiple OR equality predicates into a single IN predicte
+ * e.g COL = 1 OR COl = 2 OR COL = 3 is rewritten as COL IN (1, 2, 3)
+ *
+ * (2) single value IN predicate into equality predicate
+ * e.g COL IN (1) is rewritten as COL = 1
+ *
+ * (3) Deduplicates the predicate values for IN.
+ *
+ * These rewrites are also done for child subtrees (handling nested nature)
+ *
+ * NOTE: This optimizer is dependent on {@link FlattenNestedPredicatesFilterQueryTreeOptimizer}
+ * first flattening the tree and removing redundant operators.
  */
 public class MultipleOrEqualitiesToInClauseFilterQueryTreeOptimizer extends FilterQueryTreeOptimizer {
+
   @Override
   public FilterQueryTree optimize(FilterQueryOptimizerRequest request) {
-    return optimize(request.getFilterQueryTree(), null);
+    return optimize(request.getFilterQueryTree());
   }
 
-  private FilterQueryTree optimize(FilterQueryTree filterQueryTree, FilterQueryTree parent) {
+  /**
+   * Optimize a FilterQueryTree. When we look at the root operator,
+   * there are 3 possible cases:
+   *
+   * (1) root is OR
+   * (2) root is AND
+   * (3) root itself is a leaf level operator like =, <, >, BETWEEN, IN
+   *
+   * For (1) and (2), we look at the subtrees (children). For (3), currently
+   * the only case where rewrite is needed is for single predicate IN
+   *
+   * @param filterQueryTree root of tree to optimize/rewrite
+   * @return root of optimized tree
+   */
+  private FilterQueryTree optimize(FilterQueryTree filterQueryTree) {
     if (filterQueryTree.getOperator() == FilterOperator.OR) {
-      Map<String, Set<String>> columnToValues = new HashMap<>();
-      List<FilterQueryTree> nonEqualityOperators = new ArrayList<>();
+      // Root operator is OR
+      return optimizeTreeRootedAtOR(filterQueryTree);
+    } else if (filterQueryTree.getOperator() == FilterOperator.AND) {
+      // Root operator is AND
+      return optimizeTreeRootedAtAND(filterQueryTree);
+    } else {
+      // if we are here, it means the query has a single predicate
+      // (just one column with one or more predicate values)
+      // a = 1
+      // a IN (1)
+      // a > 7
+      // a IN (1, 2, 3, 4)
+      // .... anything single predicate
+      // rewrite single predicate IN as EQUALITY, for remaining cases this optimizer is a NO-OP
+      if (filterQueryTree.getOperator() == FilterOperator.IN && filterQueryTree.getValue().size() == 1) {
+        return new FilterQueryTree(filterQueryTree.getColumn(), filterQueryTree.getValue(), FilterOperator.EQUALITY, null);
```

Comment:
Here you also need to dedup the values for IN clause

Reply:
Done